### ISSUE TITLE
Custom loss for multiclass classification

### ISSUE THREAD

**Issue by kosnikos** [microsoft/LightGBM#6649]
I am trying to write a custom loss function for multiclass applications and I started by replicating the 'multiclassova' objective

```
def custom_classification_loss(y_true, raw_pred):
    # Reshape raw_pred to (N, num_classes)
    num_classes = 3  # Change this to your actual number of classes
    raw_pred = raw_pred.reshape(-1, num_classes)

    # Apply sigmoid for binary classification on each class (one-vs-all)
    y_pred = 1 / (1 + np.exp(-raw_pred))  # Sigmoid

    # One-hot encode the true labels
    y_true_one_hot = np.zeros_like(raw_pred)
    y_true_one_hot[np.arange(len(y_true)), y_true.astype(int)] = 1

    # Gradient for binary cross-entropy for each class
    grad = y_pred - y_true_one_hot

    # Hessian for binary cross-entropy for each class
    hess = y_pred * (1 - y_pred)
    
    return grad.ravel(), hess.ravel()
```

I am calling this function using
```lgb.LGBMClassifier(objective=custom_classification_loss, num_class= 3)```

and making predictions by applying the sigmoid function
```probabilities = sigmoid(raw_preds)```

However, the results are completely different (obviously much worse) than the predictions using the built-in 'multiclassova' objective.

Does anyone now how  'multiclassova' is implemented under the hoods?

**Comment by jmoralez** [comment]
Hey @kosnikos, thanks for using LightGBM. This is most likely because of the `init_score`, can you compare your results vs the built-in objective setting [boost_from_average](https://lightgbm.readthedocs.io/en/stable/Parameters.html#boost_from_average) to `False`? That'd make the boosting start from zero for both. Also note that when using a custom objective the predictions returned from the model are always the raw scores, so you'll have to convert them.

If you want to use the built-in init scores you can find an example on how to replicate the L2 loss in https://github.com/microsoft/LightGBM/issues/5114#issuecomment-1084994020, note that you'll have to adjust the calculation of the init scores to use the ones for `BinaryLogloss` https://github.com/microsoft/LightGBM/blob/41ba9e8f00c89d72e5cb71c964722ce1ed4d8445/src/objective/binary_objective.hpp#L139-L165

You can find the implementation for the `MulticlassOVA` objective here https://github.com/microsoft/LightGBM/blob/41ba9e8f00c89d72e5cb71c964722ce1ed4d8445/src/objective/multiclass_objective.hpp#L186-L276
